---
title: "conserved_funcional_groups_fly_yeast"
author: "Vera"
date: "24/01/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

## In this Markdown, I'm gonna try to see if there are any functional groups/clusters conserved between yeast and fly

Possible functional groups would be genes/proteins in the ribosome or other central complexes/pathways.

## 1. Load packages

```{r}

library(tidyverse)
library(dplyr)
library(ggplot2)
library(VennDiagram)
#library(ggsignif)

```

## 2. Load fly data (Florian, 29.01.2020), yeast data (Andrews lab -"thecellmap.org", 27.01.2020), and list of gene orthologues (Florian)

```{r}

# Load list of orthologue pairs from Ensembl database -27.11.2019
orthologues <- read_delim("/Users/b110/Desktop/VeraPeters/conserved_functional_groups/new_orthologues_list_copy.txt",delim = "\t")

# I also need datasets containing genetic interaction profile similarity values:

# Genetic interaction profile similarity values for Yeast:
# read genetic interaction profile similarity values (yeast, https://thecellmap.org/costanzo2016/ - downloaded Jan 27, 2020 (Data File S3))):
# description: Matrix file containing genetic interaction profile similarity values (as measured by Pearson correlation) for every pair of mutant strains in the dataset. Similarity values were computed for all gene pairs combined (ALL). The matrix contains 2 sets of row and column headers, providing a unique allele name for every mutant strain (row & column header #1) as well as a systematic ORF name (row & column header #2).
yeast_correlation <- read_delim("/Users/b110/Desktop/VeraPeters/source_data/costanzo_2016/Data File S3. Genetic interaction profile similarity matrices/cc_ALL.txt", delim = "\t")

# Genetic interaction profile similarity values for Fly:
load(file = "/Users/b110/Desktop/VeraPeters/conserved_functional_groups/cell_count_interaction_correlations_pearson_p_unfiltered_fly.RData")
# this file is named GI_corrs

# load complex annotations table (fly) - copied from Maria's folder
fly_complex_annotations <- read_delim("/Users/b110/Desktop/VeraPeters/conserved_functional_groups/complex_annotations(1).txt", delim = "\t") %>% 
  rename(Gene_name = genes)

```

## 3. Tidy up yeast data part 1 - overview and separation into gene and allele

```{r}

# I think I do not need the ORFs, so I'll drop the row/column containing the ORF names (ORF is in second column (was named X2) and in second row) and I'll rename the column with the allele name
yeast_correlation_back <- yeast_correlation
dim(yeast_correlation_back)
# 6231 6232

yeast_correlation <- yeast_correlation %>% 
  rename(allele_name = X1) %>%    #rename column for allele
  select(-X2) %>%                 #drop column X2 (= column containing ORF names)
  drop_na()                       #there is only one row containing NAs which is the first row, which has the ORF names, so I'll just use the drop_na() function to get rid of the row
dim(yeast_correlation)
# 6230 6231 <- one column and one row less than before 

# # There are duplicates (due to the fact that the matrix is symmetrical), for example, there is gene_name_1 aap1 and gene_name_2 aat2, as well as gene_name_1 aat2 and gene_name_2 aap1
# example: 
# aat2  NA  aap1  NA  -0.05033
# aap1  NA  aat2  NA  -0.05033

# set column allele_names as rownames
yeast_correlation_rownames <- yeast_correlation %>% 
  column_to_rownames(var = "allele_name")
# first, I need to check whether the column names and row names are in the same order

identical(rownames(yeast_correlation_rownames), colnames(yeast_correlation_rownames))
# [1] TRUE        # fortunately, they are

# replace upper triangle of the matrix NA, including the diagonale
yeast_correlation_rownames[upper.tri(yeast_correlation_rownames, diag=TRUE)] <- NA 

# convert rownames back to column
yeast_correlation <- yeast_correlation_rownames%>% 
  rownames_to_column(var = "allele_name")

# gather
yeast_correlation_gathered <- yeast_correlation %>% 
  gather(allele_name_2, value, -allele_name)
dim(yeast_correlation_gathered)
#[1] 38812900        3

# drop the NAs. Additionally, for some allele combinations, the value is "NaN"" - I cannot use those, so I'll filter them out
yeast_correlation_gathered <- yeast_correlation_gathered %>% 
  drop_na() %>% 
  filter(value != "NaN", value != "NA")     #for whatever reason, the drop_na() did not work, but filtering for the dtring "NA" did
dim(yeast_correlation_gathered)
# [1] 17516577        3 
# less than half of the rows. makes sense, because diagonal is excluded

# get an overview of the similarity values:
min(as.numeric(yeast_correlation_gathered$value))   # !!! without the "as.numeric", value is incorrect, gives -0.0000 instead
# [1] -0.38475
max(as.numeric(yeast_correlation_gathered$value))
# [1] 0.82245
mean(as.numeric(yeast_correlation_gathered$value))
# [1] 0.006681768

# Let's see how many similarity values are >0.2 ( similarity value of 0.2 was used by Costanzo et al. as threshold to create genetic interaction networks)
nrow(yeast_correlation_gathered %>% filter(value > 0.2))
# 39857 (slightly less than I had when filtering the dataset which still included duplicates (79732))
# rerun 30.01.: 39866

# separate allele_name into gene name and allele name:
# Problem: some of the allele names have two "-" in them, so it would be separated into three total columns -> I need to find an opportunity to separate only at the first "-"
# Let's try this (merge allows only as many separations as number of new columns I give):
yeast_correlation_gathered_sep <- yeast_correlation_gathered %>% 
  separate(allele_name, into = c("gene_name_1", "allele_name_1"), sep = "-", extra = "merge") %>% 
  separate(allele_name_2, into = c("gene_name_2", "allele_name_2"), sep = "-", extra = "merge")

```

## 4. Tidy up yeast data part 2 - only similarity values of genes present in the orthologues list

```{r}

# how many unique gene names are there which are present in the orthologues table?
# I think I have to use the S_cerevisiae_gene_name column here instead of S_cerevisiae_stable_gene_ID
yeast_correlation_gene_names_1 <- yeast_correlation_gathered_sep %>% 
  select(gene_name_1) %>%
  unique() 
dim(yeast_correlation_gene_names_1)
# [1] 5691    1

yeast_correlation_gene_names_2 <- yeast_correlation_gathered_sep %>% 
  select(gene_name_2) %>%
  unique() 
dim(yeast_correlation_gene_names_2)
# [1] 5691    1

which(! yeast_correlation_gene_names_1$gene_name_1 %in% yeast_correlation_gene_names_2$gene_name_2)
# [1] 5531
# the gene names in the two columns are different (makes sense because I deleted half of the matrix ?)

# combine both subsets, filter unique gene names
yeast_correlation_gene_names_1 <- rename(yeast_correlation_gene_names_1, "gene_name" = "gene_name_1")
yeast_correlation_gene_names_2 <- rename(yeast_correlation_gene_names_2, "gene_name" = "gene_name_2")

yeast_correlation_gene_names <- rbind(yeast_correlation_gene_names_1, yeast_correlation_gene_names_2) %>% 
  unique()
dim(yeast_correlation_gene_names)
# [1] 5692    1   # so, this procedure saved me one gene name! :D

# Problem: the gene names in the yeast_correlation_gene_names dataset are written in lowercase letters, while those in orthologues are written in uppercase letters -> I need to rewrite one of them or ignore the case, in order to be able to filter 

# create a vector containing all unique S_cerevisiae_gene_names in the orthologues dataframe, in lowercase letters
orthologues_unique <- orthologues %>% 
  select(S_cerevisiae_gene_name) %>% 
  drop_na() %>% 
  unique()
orthologues_gene_names <- orthologues_unique[["S_cerevisiae_gene_name"]] %>% 
  tolower()
# yay, a character vector with lowercase gene names
length(orthologues_gene_names)
# [1] 1186

# Now, let's check which/how many gene names are present in our dataset
yeast_correlation_gene_names_in_ortho <- yeast_correlation_gene_names %>% 
  filter(gene_name %in% orthologues_gene_names)
dim(yeast_correlation_gene_names_in_ortho)
# [1] 1056    1
# from the originally 5692 unique gene names, only 1056 are left after filtering with the orthologues gene names 

# check, whether gene_stable_ID was also used as gene name in the yeast dataset:
orthologues_stableid <- orthologues %>% 
  select(S_cerevisiae_gene_stable_ID) %>% 
  drop_na() %>% 
  unique()
orthologues_stableid_vec <- orthologues_stableid[["S_cerevisiae_stable_ID"]] %>% 
  tolower()
which(yeast_correlation_gathered_sep$gene_name_2 %in% orthologues_stableid_vec)
# integer(0) # stable ID was not used

# Let's create a new dataframe, containing only rows, whose gene_name_1 and gene_name_2 are present in the orthologues list
yeast_correlation_in_ortho <- yeast_correlation_gathered_sep %>% 
  filter(gene_name_1 %in% orthologues_gene_names & gene_name_2 %in% orthologues_gene_names)
dim(yeast_correlation_in_ortho)
#(# [1] 1551914       5   # from the 35 million, only 1.5 million rows are left)
# [1] 775486      5     after filtering duplicates

```

## 5. Density histogram of similarity values of yeast genes present in orthologues dataset

```{r}

# Plot density histogram of similarity values of 
ggplot(yeast_correlation_in_ortho, aes(as.numeric(value))) +
  theme_bw() +
  geom_histogram(binwidth = 0.001, color = "darkgoldenrod") +
  xlab("similarity value") +
  geom_vline(xintercept = 0, color = "red") +
  ggtitle(" Density histogram of similarity values of yeast genes present \n in the orthologues list") +
  scale_x_continuous(breaks = c(-0.2, 0, 0.2, 0.4, 0.6, 0.8))
ggsave(file="/Users/b110/Desktop/VeraPeters/conserved_functional_groups/densityhist_similarityvalues_yeast_in_ortho.png")

max(as.numeric(yeast_correlation_in_ortho$value)) 
min(as.numeric(yeast_correlation_in_ortho$value)) 
 
# just out of interest, let's plot the density histogram before filtering for genes present in orthologues dataset:
ggplot(yeast_correlation_gathered, aes(as.numeric(value))) +
  theme_bw() +
  geom_histogram(binwidth = 0.001, color = "darkgoldenrod") +
  xlab("pearson correlation") +
  geom_vline(xintercept = 0, color = "red") +
  ggtitle(" Density histogram of pearson correlation values, yeast ") +
  scale_x_continuous(breaks = c(-0.2, 0, 0.2, 0.4, 0.6, 0.8))
ggsave(file="/Users/b110/Desktop/VeraPeters/conserved_functional_groups/densityhist_similarityvalues_yeast_unfiltered.png")

median(as.numeric(yeast_correlation_in_ortho$value))
#[1] 0.00476
median(as.numeric(yeast_correlation_gathered$value))
#[1] 0.00492

```

## 6. Compare filtered yeast genes (present in orthologues) to fly complexes 

```{r}

# check dimensions of complex annotations
dim(fly_complex_annotations)
# [1] 547   2

# use complex annotations in new column of orthologues, keep only rows which have genes with matching complex annotation (inner_join)
orthologues_annotations <- inner_join(orthologues, fly_complex_annotations, by = "Gene_name")
nrow(orthologues_annotations)
# [1] 1020

# create vectors out of the yeast gene names in the yeast_correlation_in_ortho
yeast_gene_names_in_ortho_1 <- yeast_correlation_in_ortho[["gene_name_1"]] %>% 
  toupper()
yeast_gene_names_in_ortho_2 <- yeast_correlation_in_ortho[["gene_name_2"]] %>% 
  toupper()

# filter orthologues_annotations for yeast genes present in the similarity values, keep only columns I need
orthologues_annotations_yeast_sim_value <- orthologues_annotations %>% 
  filter(S_cerevisiae_gene_name %in% yeast_gene_names_in_ortho_1 & S_cerevisiae_gene_name %in% yeast_gene_names_in_ortho_2) %>% 
  select(Gene_stable_ID, S_cerevisiae_gene_stable_ID, S_cerevisiae_gene_name, S_cerevisiae_orthology_confidence_0low_1high, percentage_id_query_gene_identical_to_target_S_cerevisiae_gene, Gene_name, complex) %>% 
  distinct()
nrow(orthologues_annotations_yeast_sim_value)
# [1] 187        

# let's see which (fly) complexes there are:
fly_complexes_in_yeast_sim_value <- orthologues_annotations_yeast_sim_value %>% 
  select(complex) %>% 
  unique()
fly_complexes_in_yeast_sim_value_vec <- fly_complexes_in_yeast_sim_value[["complex"]]
length(fly_complexes_in_yeast_sim_value_vec)
# [1] 27          # yeast genes corresponding to 27 fly complexes are present in the similarity values matrix
fly_complexes_in_yeast_sim_value_vec
# [1] "RAS"     "OXPHOS"  "ESCRT"   "COPII"   "PROT"    "HSP60"   "PATPASE" "APC"     "AP"      "TIP60"   "RETRO"   "IlRNeg"  "TFIID"   "VPSC"    "IlRPos"  "COMP"  
# [17] "RNAPII"  "COP9"    "IlR"     "BAPPBAP" "EXOC"    "COPI"    "PFDN"    "ARPC"    "MED"     "GATOR"   "TRAPP"  

# let's see how often the complex is present
orthologues_annotations_yeast_sim_value %>% 
  select(Gene_stable_ID, S_cerevisiae_gene_stable_ID, S_cerevisiae_gene_name, S_cerevisiae_orthology_confidence_0low_1high, percentage_id_query_gene_identical_to_target_S_cerevisiae_gene, Gene_name, complex) %>% 
  distinct() %>% 
  group_by(complex) %>% 
  count() %>% 
  view() #%>%
  #write.table(file = "/Users/b110/Desktop/VeraPeters/conserved_functional_groups/complexes_count_in_yeast_sim_values.txt", sep = "\t")

```

## 7. Tidy up fly data

```{r}

# the fly dataset already has the following modifications: Pearson correlation coefficient was computed and the correlation matrix was generated, the upper triangle was set to NA, matrix was gathered and NAs were removed

# use FLorian's code to separate gene and IDs into single columns:
fly_correlation_gathered_sep <- GI_corrs %>%
  separate(x,c("genex","idx"),sep = "öö")%>%
  separate(y,c("geney","idy"),sep = "öö") %>%
  dplyr::select(genex,geney,idx,idy,r) 
dim(fly_correlation_gathered_sep)
# [1] 23526370        5

# some of the correlations correspond to wells without dsRNA or with only one dsRNA?
# in those columns, the gene is marked with "-" -> I'm gonna filter out lines in which genex or geney equals "-"
fly_correlation_gathered_sep <- fly_correlation_gathered_sep %>% 
  filter(genex != "-", geney != "-")
dim(fly_correlation_gathered_sep)
# [1] 23498940        5         # 27430 rows less

# Let's first filter the dataframe, so it contains only fly genes which are present in the orthologues list
fly_correlation_in_ortho <- fly_correlation_gathered_sep %>%
  filter(genex %in% orthologues$Gene_name & geney %in% orthologues$Gene_name)
nrow(fly_correlation_in_ortho)
# [1] 22481865                  # most of the fly correlation genes are in the orthologues list (still, ca. 1 million rows got lost in this step)

# Now, filter orthologues_annotation table to see which complexes have genes represented in the fly_correlation dataset, which complexes they are and how many elements per complex group
orthologues_annotations_fly_sim_value <- orthologues_annotations %>% 
  filter(Gene_name %in% fly_correlation_in_ortho$genex & Gene_name %in% fly_correlation_in_ortho$geney) %>% 
  select(Gene_stable_ID, S_cerevisiae_gene_stable_ID, S_cerevisiae_gene_name, S_cerevisiae_orthology_confidence_0low_1high, percentage_id_query_gene_identical_to_target_S_cerevisiae_gene, Gene_name, complex) %>% 
  distinct()
nrow(orthologues_annotations_fly_sim_value)
# [1] 393

# let's see which (fly) complexes there are:
fly_complexes_in_fly_sim_value <- orthologues_annotations_fly_sim_value %>% 
  select(complex) %>% 
  unique()
fly_complexes_in_fly_sim_value_vec <- fly_complexes_in_fly_sim_value[["complex"]]
length(fly_complexes_in_fly_sim_value_vec)
# [1] 32
fly_complexes_in_fly_sim_value_vec
# [1] "IlRNeg"   "DYN"      "MED"      "ESCRT"    "COPII"    "CLATHRIN" "IlR"      "OXPHOS"   "PROT"     "HSP60"    "PATPASE"  "TFIID"    "RAS"      "APC"      "COPI"  
# [16] "TIP60"    "IFT"      "ARPC"     "BAPPBAP"  "TRAPP"    "GATOR"    "IlRPos"   "AP"       "VPSC"     "EXOC"     "COMP"     "STAT"     "RNAPII"   "COP9"     "PFDN" 
# [31] "RETRO"    "ELG1C" 

# let's see how often the complex is present
orthologues_annotations_fly_sim_value %>% 
  select(Gene_stable_ID, S_cerevisiae_gene_stable_ID, S_cerevisiae_gene_name, S_cerevisiae_orthology_confidence_0low_1high, percentage_id_query_gene_identical_to_target_S_cerevisiae_gene, Gene_name, complex) %>% 
  distinct() %>% 
  group_by(complex) %>% 
  count() %>% 
  view() %>%
  write.table(file = "/Users/b110/Desktop/VeraPeters/conserved_functional_groups/complexes_count_in_fly_sim_values.txt", sep = "\t")

```

## 8. Density histograms of similarity values of fly genes present in orthologues dataset

```{r}

# Plot density histogram of similarity values of 
ggplot(fly_correlation_in_ortho, aes(r)) +
  theme_bw() +
  geom_histogram(binwidth = 0.001, color = "blue3") +
  xlab("similarity value") +
  geom_vline(xintercept = 0, color = "red") +
  ggtitle(" Density histogram of similarity values of fly genes present \n in the orthologues list")
ggsave(file="/Users/b110/Desktop/VeraPeters/conserved_functional_groups/densityhist_similarityvalues_fly_in_ortho.png")
 
# just out of interest, let's plot the density histogram before filtering for genes present in orthologues dataset:
ggplot(GI_corrs, aes(r)) +
  theme_bw() +
  geom_histogram(binwidth = 0.001, color = "blue3") +
  xlab("similarity value") +
  geom_vline(xintercept = 0, color = "red") +
  ggtitle(" Density histogram of all fly similarity values ")
ggsave(file="/Users/b110/Desktop/VeraPeters/conserved_functional_groups/densityhist_similarityvalues_fly_unfiltered.png")

median(fly_correlation_in_ortho$r)
# 0.002213179
median(GI_corrs$r)
# 0.002262668

```

## 9. Compare the complexes, filter for complex group size >4

```{r}

which(! fly_complexes_in_yeast_sim_value_vec %in% fly_complexes_in_fly_sim_value_vec)
# integer(0)              # all the complexes that have gene members in the yeast similarity values also have members in the fly simlarity values
which(! fly_complexes_in_fly_sim_value_vec %in% fly_complexes_in_yeast_sim_value_vec)
# [1]  2  6 17 27 32      # in the fly similarity values, there are genes that belong to 5 complexes, which are not in the yeast similarity values

# filter complex group size >4
# yeast dataset
yeast_complexes_n4 <- orthologues_annotations_yeast_sim_value %>% 
  select(Gene_stable_ID, S_cerevisiae_gene_stable_ID, S_cerevisiae_gene_name, S_cerevisiae_orthology_confidence_0low_1high, percentage_id_query_gene_identical_to_target_S_cerevisiae_gene, Gene_name, complex) %>% 
  distinct() %>% 
  group_by(complex) %>% 
  filter(n() >= 4) %>% 
  count() 
nrow(yeast_complexes_n4)
# [1] 22

# fly dataset:
fly_complexes_n4 <- orthologues_annotations_fly_sim_value %>% 
  select(Gene_stable_ID, S_cerevisiae_gene_stable_ID, S_cerevisiae_gene_name, S_cerevisiae_orthology_confidence_0low_1high, percentage_id_query_gene_identical_to_target_S_cerevisiae_gene, Gene_name, complex) %>% 
  distinct() %>% 
  group_by(complex) %>% 
  filter(n() >= 4) %>% 
  count()
nrow(fly_complexes_n4)
# [1] 27

which(! fly_complexes_n4$complex %in% yeast_complexes_n4$complex)
# [1]  6  8 10 13 16 23 26        # there are 7 fly complexes, which have n >= 4 that do not have n >= 4 in yeast
which(! yeast_complexes_n4$complex %in% fly_complexes_n4$complex)
# [1] 15 18                       # there are 2 yeast complexes, which have n >= 4 that do not have n >= 4 in fly

# create common dataframe with the counts 
yeast_complexes_n4 <- yeast_complexes_n4 %>% rename(count_yeast = n)
fly_complexes_n4 <- fly_complexes_n4 %>% rename(count_fly = n)

complexes_n4 <- full_join(yeast_complexes_n4, fly_complexes_n4, by = "complex") %>% 
  drop_na()
write.table(complexes_n4, file = "/Users/b110/Desktop/VeraPeters/conserved_functional_groups/complexes_n4.txt", sep = "\t")
nrow(complexes_n4)
# 20        # I have 20 complexes left, that have at least 4 gene members in both datasets
complexes_n4_vec <- complexes_n4[["complex"]]

# create Venn diagram to visualize
grid.newpage()
draw.pairwise.venn(area1 = nrow(fly_complexes_n4), area2 = nrow(yeast_complexes_n4), cross.area = nrow(complexes_n4), category = c("Fly", "Yeast"), fill = c("blue1", "darkgoldenrod"), scaled=FALSE, cat.dist = c(0.05), fontface = c(2), lwd = c(1), alpha = c(0.35), cex = c(1.5), cat.cex = c(1.5), cat.fontface = c(2), cat.pos = c(-30,30))
grid.text("Complexes with with at least 4 gene pairs", vjust = -23, gp = gpar(fontsize = 15, fontface = "bold"))
grid.text("total: Fly: 27, Yeast: 22", vjust = 28)
dev.copy2pdf(file="/Users/b110/Desktop/VeraPeters/conserved_functional_groups/venndiagram_complexes_n4.pdf")

```

## 10. Filter the datasets of yeast and fly - keep only rows containing gene pairs corresponding to complexes with n >= 4 gene pairs in both datasets

```{r}

# datasets to filter: yeast_correlation_gathered_sep, fly_correlation_gathered_sep

# first, filter orthologues_annotations table to keep only the complexes with n >= 4 gene pairs in both datasets
orthologues_annotations_n4 <- orthologues_annotations %>% 
  select(Gene_stable_ID, S_cerevisiae_gene_stable_ID, S_cerevisiae_gene_name, S_cerevisiae_orthology_confidence_0low_1high, percentage_id_query_gene_identical_to_target_S_cerevisiae_gene, Gene_name, complex) %>% 
  filter(complex %in% complexes_n4$complex) %>% 
  distinct()
nrow(orthologues_annotations_n4)
# [1] 486

# Problem?: there are Gene_names (fly) which do not have a corresponding S_cerevisiae_gene_name (they are NA)
# but: I wanna look at complexes, the single gene does not necessarily have to be present in both organisms (I think)

# to filter the yeast dataset, I have to create a vector again, because of the uppercase/lowercase issue (and remove NA values from this vector)
yeast_gene_names_complexes_n4 <- orthologues_annotations_n4[["S_cerevisiae_gene_name"]] %>% 
  tolower() %>% 
  discard(is.na)

# filter yeast dataset:
yeast_correlation_gathered_sep_n4 <- yeast_correlation_gathered_sep %>% 
  filter(gene_name_1 %in% yeast_gene_names_complexes_n4 & gene_name_2 %in% yeast_gene_names_complexes_n4)
nrow(yeast_correlation_gathered_sep_n4)
# [1] 17181

# filter fly dataset:
fly_correlation_gathered_sep_n4 <- fly_correlation_gathered_sep %>% 
  filter(genex %in% orthologues_annotations_n4$Gene_name & geney %in% orthologues_annotations_n4$Gene_name)
nrow(fly_correlation_gathered_sep_n4)
# [1] 46360

# I have some gene pairs in these datasets, whose members belong to different complex annotations. for example gene 1 to complex 1, gene 2 to complex 2, these would be external interactions (both genes belonging to the same complex would be internal interactions)

# Now it might make sense to add the columns complex 1 (gene 1) and complex 2 (gene 2) to those filtered datasets, I should be able to use a left_join with the orthologues list for this; later I could filter internal interactions by filtering for rows where complex 1 = complex 2 (this means that both genes are in the same complex)

# to avoid renaming every time, I'm gonna modify the orthologues dataset, copy some columns and rename them
orthologues_annotations_n4_compare <- orthologues_annotations_n4 %>% 
  select(S_cerevisiae_gene_name, Gene_name, complex) %>% 
  rename(gene_name_1 = S_cerevisiae_gene_name, genex = Gene_name) %>% 
  mutate(gene_name_2 = gene_name_1, geney = genex) %>% 
  select(gene_name_1, gene_name_2, genex, geney, complex) %>% 
  mutate(gene_name_1 = tolower(gene_name_1), gene_name_2 = tolower(gene_name_2))

# create complex columns for yeast dataset, add column for internal/external interaction
yeast_correlation_gathered_sep_n4_complex <- left_join(yeast_correlation_gathered_sep_n4, orthologues_annotations_n4_compare, by = "gene_name_1") %>% 
  select(-gene_name_2.y, -genex, -geney) %>% 
  rename(gene_name_2 = gene_name_2.x, complex_1 = complex) %>% 
  left_join(orthologues_annotations_n4_compare, by = "gene_name_2") %>% 
  select(-gene_name_1.y, -genex, -geney) %>% 
  rename(gene_name_1 = gene_name_1.x, complex_2 = complex) %>% 
  mutate(type = ifelse(complex_1 == complex_2, "internal", "external")) %>% 
  distinct()      # for whatever reason, there are duplicated rows, which  I do not wanna have cause the duplicated values create "wrong" distributions and pvalues

# create complex columns for fly dataset, add column for internal/external interaction
fly_correlation_gathered_sep_n4_complex <- left_join(fly_correlation_gathered_sep_n4, orthologues_annotations_n4_compare, by = "genex") %>% 
  select(-gene_name_1, -gene_name_2, -geney.y) %>% 
  rename(geney = geney.x, complex_x = complex) %>% 
  left_join(orthologues_annotations_n4_compare, by = "geney") %>% 
  select(-gene_name_1, -gene_name_2, -genex.y) %>% 
  rename(genex = genex.x, complex_y = complex) %>% 
  mutate(type = ifelse(complex_x == complex_y, "internal", "external")) %>% 
  distinct()      # for whatever reason, there are duplicated rows, which  I do not wanna have cause the duplicated values create "wrong" distributions and pvalues

# making sure, that every complex is present in both datasets fo both genes:
yeast_correlation_gathered_sep_n4_complex %>% select(complex_1) %>% unique() %>% count()
# 20
yeast_correlation_gathered_sep_n4_complex %>% select(complex_2) %>% unique() %>% count()
# 20
fly_correlation_gathered_sep_n4_complex %>% select(complex_x) %>% unique() %>% count()
# 20
fly_correlation_gathered_sep_n4_complex %>% select(complex_y) %>% unique() %>% count()
# 20
# hooray

```

## 11. Plot Densities of distributions internal vs external similarity correlations

```{r}

# I'm thinking of plotting one plot per complex; each plot has two subplots (one per organism); each subplots consists of to geom_density (one representing the internal, one the external interaction profile similarity correlation distribution)

# Testplot for PROT (proteasome):
# fly:
fly_correlation_gathered_sep_n4_complex %>% 
  filter(complex_x == "PROT" | complex_y == "PROT") %>% 
  ggplot(aes(x = r, fill = type)) +
    geom_density(alpha = 0.3) +
    scale_fill_manual(values = c("grey", "blue3")) +
    theme_bw() +
    xlab("similarity value") +
    ylab("count") +
    geom_vline(xintercept = 0, color = "black") +
    ggtitle("Density histogram of similarity values for complex 'PROT' in fly")
ggsave(file="/Users/b110/Desktop/VeraPeters/conserved_functional_groups/density_hist_prot_fly_test.png")

# yeast:
yeast_correlation_gathered_sep_n4_complex %>% 
  filter(complex_1 == "PROT" | complex_2 == "PROT") %>% #view()
  ggplot(aes(x = as.numeric(value), fill = type)) +
    geom_density(alpha = 0.3) +
    scale_fill_manual(values = c("grey", "darkgoldenrod")) +
    theme_bw() +
    xlab("similarity value") +
    ylab("count") +
    geom_vline(xintercept = 0, color = "black") +
    ggtitle("Density histogram of similarity values for complex 'PROT' in yeast")
ggsave(file="/Users/b110/Desktop/VeraPeters/conserved_functional_groups/density_hist_prot_yeast_test.png")

# Let's try doing this in a for loop for each of the 20 complexes:

# create common dataset for yeast and fly
# first: rename columns in the datasets
fly_correlation_gathered_sep_n4_complex_new <- fly_correlation_gathered_sep_n4_complex %>% 
  select(-idx, -idy) %>% 
  rename(Gene_1 = genex, Gene_2 = geney, sim_val = r, Complex_1 = complex_x, Complex_2 = complex_y) %>% 
  mutate(sim_val = as.character(sim_val))

yeast_correlation_gathered_sep_n4_complex_new <- yeast_correlation_gathered_sep_n4_complex %>% 
  select(-allele_name_1, -allele_name_2) %>% 
  rename(Gene_1 = gene_name_1, Gene_2 = gene_name_2, sim_val = value, Complex_1 = complex_1, Complex_2 = complex_2)

fly_yeast_correlation_complex <- bind_rows(fly_correlation_gathered_sep_n4_complex_new, yeast_correlation_gathered_sep_n4_complex_new, .id = "organism")
fly_yeast_correlation_complex$organism[fly_yeast_correlation_complex$organism == 1] <- "fly"
fly_yeast_correlation_complex$organism[fly_yeast_correlation_complex$organism == 2] <- "yeast"

# include pvalues:
# pvalue calculation:
# create new empty dataframes to write pvalues in:

tt_fly <- data.frame(matrix(ncol = 3, nrow = length(complexes_n4_vec))) %>% 
  rename(organism = X1, complex = X2, pval = X3)

tt_yeast <- data.frame(matrix(ncol = 3, nrow = length(complexes_n4_vec))) %>% 
  rename(organism = X1, complex = X2, pval = X3)

# fly
j <- 1
for (i in (complexes_n4_vec)) {
    tt_fly$organism[j] <- "fly"  
    tt_fly$complex[j] <- as.character(i)
    tt_fly$pval[j] <- t.test(formula = as.numeric(sim_val) ~type, alternative = c("two.sided"), data = fly_yeast_correlation_complex %>% filter((Complex_1 == i | Complex_2 == i), organism == "fly"))$p.value
    j <- j+1
}     

#yeast
j <- 1
for (i in (complexes_n4_vec)) {
    tt_yeast$organism[j] <- "yeast"  
    tt_yeast$complex[j] <- as.character(i)
    tt_yeast$pval[j] <- t.test(formula = as.numeric(sim_val) ~type, alternative = c("two.sided"), data = fly_yeast_correlation_complex %>% filter((Complex_1 == i | Complex_2 == i), organism == "yeast"))$p.value
    j <- j+1
} 

# combine both dataframes in one: 
tt <- as_tibble(bind_rows(tt_fly, tt_yeast)) %>% 
  mutate(pvalround = ifelse(pval < 0.00001, "< 0.00001", as.character(paste0 ("= ",round(pval, digits = 5)))))  # round pvalue/set to < x, to make display easier

# plot:
# for loop: one plot per complex (fly):
for (i in (complexes_n4_vec)) {
plt <- fly_yeast_correlation_complex %>% 
  filter(organism == "fly") %>% 
  filter(Complex_1 == i | Complex_2 == i) %>% 
  ggplot(aes(x = as.numeric(sim_val), fill = type)) +
    geom_density(alpha = 0.3) +
    scale_fill_manual(values = c("grey", "blue3")) +
    theme_classic(base_size = 15) +
    xlab("similarity value") +
    ylab("count") +
    geom_vline(xintercept = 0, color = "black", linetype ="dashed") +
    scale_y_continuous(expand = c(0,0)) +
    ggtitle(paste0("Distribution of similarity values, complex ", as.character(i)))

plt +
    geom_text(aes(label = paste0("p ", (tt %>%  filter(complex == i, organism == "fly"))$pvalround)), x = layer_scales(plt)$x$range$range[2]-(layer_scales(plt)$x$range$range[2]/5), y = layer_scales(plt)$y$range$range[2]-layer_scales(plt)$y$range$range[2]/10, size = 5)
 
ggsave(paste0("/Users/b110/Desktop/VeraPeters/conserved_functional_groups/Plots/density_hist_fly_", i, ".png"))
}

# for the special plots in fly... (distributions were weird, so that I had to adjust the position of geom_text):
# RNAPII
plt <- fly_yeast_correlation_complex %>% 
  filter(organism == "fly") %>% 
  filter(Complex_1 == "RNAPII" | Complex_2 == "RNAPII") %>% #view()
  ggplot(aes(x = as.numeric(sim_val), fill = type)) +
    geom_density(alpha = 0.3) +
    scale_fill_manual(values = c("grey", "blue3")) +
    theme_classic(base_size = 15) +
    xlab("similarity value") +
    ylab("count") +
    geom_vline(xintercept = 0, color = "black", linetype = "dashed") +
    ggtitle("Distribution of similarity values, complex RNAPII") +
    scale_y_continuous(expand = c(0,0))

plt +
    geom_text(aes(label = paste0("p ", (tt %>%  filter(complex =="RNAPII", organism == "fly"))$pvalround)), x = layer_scales(plt)$x$range$range[2]-(layer_scales(plt)$x$range$range[2]/2), y = layer_scales(plt)$y$range$range[2]-layer_scales(plt)$y$range$range[2]/10, size = 5)

ggsave("/Users/b110/Desktop/VeraPeters/conserved_functional_groups/Plots/density_hist_fly_RNAPII.png")

# COP9
plt <- fly_yeast_correlation_complex %>% 
  filter(organism == "fly") %>% 
  filter(Complex_1 == "COP9" | Complex_2 == "COP9") %>% #view()
  ggplot(aes(x = as.numeric(sim_val), fill = type)) +
    geom_density(alpha = 0.3) +
    scale_fill_manual(values = c("grey", "blue3")) +
    theme_classic(base_size = 15) +
    xlab("similarity value") +
    ylab("count") +
    geom_vline(xintercept = 0, color = "black", linetype = "dashed") +
    ggtitle("Distribution of similarity values, complex COP9") +
    scale_y_continuous(expand = c(0,0))

plt +
    geom_text(aes(label = paste0("p ", (tt %>%  filter(complex =="COP9", organism == "fly"))$pvalround)), x = layer_scales(plt)$x$range$range[2]-(layer_scales(plt)$x$range$range[2]/2), y = layer_scales(plt)$y$range$range[2]-layer_scales(plt)$y$range$range[2]/10, size = 5)

ggsave("/Users/b110/Desktop/VeraPeters/conserved_functional_groups/Plots/density_hist_fly_COP9.png")

# for loop: one plot per complex (yeast):
for (i in (complexes_n4_vec)) {
plt <- fly_yeast_correlation_complex %>% 
  filter(organism == "yeast") %>% 
  filter(Complex_1 == i | Complex_2 == i) %>% 
  ggplot(aes(x = as.numeric(sim_val), fill = type)) +
    geom_density(alpha = 0.3) +
    scale_fill_manual(values = c("grey", "darkgoldenrod")) +
    theme_classic(base_size = 15) +
    xlab("pearson correlation") +
    ylab("count") +
    geom_vline(xintercept = 0, color = "black", linetype ="dashed") +
    scale_y_continuous(expand = c(0,0)) +
    ggtitle(paste0(" Distribution of pearson correlations, complex ", as.character(i), ", \n yeast"))

plt +
    geom_text(aes(label = paste0("p ", (tt %>%  filter(complex == i, organism == "yeast"))$pvalround)), x = layer_scales(plt)$x$range$range[2]-(layer_scales(plt)$x$range$range[2]/5), y = layer_scales(plt)$y$range$range[2]-layer_scales(plt)$y$range$range[2]/10, size = 5)
 
ggsave(paste0("/Users/b110/Desktop/VeraPeters/conserved_functional_groups/Plots/density_hist_yeast_", i, ".png"))
}

```

## 12. Group complexes based on the shapes of their similarity value distributions, pvalues and similarity between fly and yeast

```{r}

# when looking at the plots created in chunk 11, one can see that for some complexes, the right-shift of internal similarity values (compared to external) is more profound, some distributions have two peaks, some have even a left shift, and there are complexes which display one of these options in fly, and another one in yeast. I wanna create a dataframe containing the complexes, with annotations abput their behaviour.

# To do this, I'll modify the complexes_n4 dataframe and add the calculated pvalues from tt

complexes_n4_groups <- complexes_n4 %>% 
  mutate(group = ifelse (complex %in% c("AP", "BAPPBAP"), "two_peaks",
         ifelse(complex %in% c("APC", "ESCRT", "MED", "OXPHOS", "TFIID"), "right_shift",
                ifelse(complex %in% c("GATOR", "RAS"), "left_shift",
                                ifelse(complex %in% c("COMP", "COP9", "COPII", "HSP60", "PROT", "RNAPII", "VPSC"), "differing",
                                       ifelse(complex %in% c("PATPASE", "IlR", "TIP60", "IlRNeg"), NA
          )))))) %>% 
  left_join ((tt %>% select(-pvalround) %>% spread(organism, pval) %>% rename(complex = Complex_1)), complexes_n4_groups, by = "complex") %>% 
  rename("pval_fly" = "fly", "pval_yeast" = "yeast") %>% 
  select(complex, count_fly, count_yeast, pval_fly, pval_yeast, group)    # just to re-order the columns

```

## 13. Investigate genes leading to differing similarity value distribution when comparing fly and yeast, two-peak distribution

```{r}

# -> which gene pairs are the ones in the left side of the two-peaks distribution?
# -> which gene pairs lead to distributions differing between fly and yeast?

# use groups I made in complexes_n4_groups, filter for group of interest to get complexes of interest
# filter datasets for complexes of interest
# for example for two-peak: find out which genes are internal and have similarity values around 0, compared to the ones > 0.2 or > 0.5

# use fly_yeast_correlation_complex dataframe
# colnames: "organism"  "Gene_1"    "Gene_2"    "sim_val"   "Complex_1" "Complex_2" "type"

# 1 #
# I wanna first check the complex PROT: in yeast, there is a right-shift, whereas in fly there are two peaks
# which are the genes in the left peak (fly) and what is their sequence homology with the yeast genes? If sequence high sequence homology -> functional not homologue!
# the "cut" between the peaks in the fly distribution is at ca. 0.3

genes_03_prot_fly <- fly_yeast_correlation_complex %>% 
  filter(Complex_1 == "PROT" & Complex_2 =="PROT") %>%      # I wanna filter both complexes to be "PROT", cause I'm interested in the genes of the left peak in the internal distribution
  filter(organism == "fly") %>% 
  filter(sim_val < 0.3)

# now, let' see in the orthologues_annotations_n4 table:
# what is the mean percentage id of all genes annotated woth the PROT complex vs. the genes filtered in genes_03_prot_fly

# for all genes in complex PROT:
mean((orthologues_annotations_n4 %>% 
        filter(complex == "PROT"))$percentage_id_query_gene_identical_to_target_S_cerevisiae_gene, na.rm = TRUE)
# [1] 45.73026

# for the genes in the left peak:
mean((orthologues_annotations_n4 %>% 
        filter(complex == "PROT") %>% 
        filter(Gene_name %in% genes_03_prot_fly$Gene_1 | Gene_name %in% genes_03_prot_fly$Gene_2))$percentage_id_query_gene_identical_to_target_S_cerevisiae_gene, na.rm = TRUE)
# [1] 44.48441

# in this case, the mean sequence identity is almost the same, comparing the genes in the left peak vs. all genes

# 2 #
# next, I wanna see whether the sequence identity is differing when I compare for complexes, in which the distributions show a right shift in both 
# I am choosig the complex "MED", because it looked similar in both organisms

genes_med_fly <- fly_yeast_correlation_complex %>% 
  filter(Complex_1 == "MED" & Complex_2 =="MED") %>%      # filter for internal genes
  filter(organism == "fly")                               # filter for organism fly

genes_med_yeast <- fly_yeast_correlation_complex %>%
  filter(Complex_1 == "MED" & Complex_2 =="MED") %>%      # filter for internal genes
  filter(organism == "yeast") %>%                         # filter for organism fly
  mutate(Gene_1 = toupper(Gene_1), Gene_2 = toupper(Gene_2))
  
  
# mean sequence identity for all genes in complex "MED":
mean((orthologues_annotations_n4 %>% 
        filter(complex == "MED"))$percentage_id_query_gene_identical_to_target_S_cerevisiae_gene, na.rm = TRUE)
# [1] 22.99305

# mean sequence identity for internal genes in complex "MED" (based on fly genes):
mean((orthologues_annotations_n4 %>% 
        filter(complex == "MED") %>% 
        filter(Gene_name %in% genes_med_fly$Gene_1 | Gene_name %in% genes_med_fly$Gene_2))$percentage_id_query_gene_identical_to_target_S_cerevisiae_gene, na.rm = TRUE)
# [1] 24.04565

# mean sequence identity for internal genes in complex "MED" (based on yeast genes):
mean((orthologues_annotations_n4 %>% 
        filter(complex == "MED") %>% 
        filter(S_cerevisiae_gene_name %in% genes_med_yeast$Gene_1 | S_cerevisiae_gene_name %in% genes_med_yeast$Gene_2))$percentage_id_query_gene_identical_to_target_S_cerevisiae_gene, na.rm = TRUE)
# [1] 22.8756


```








