---
title: "internal_vs_external_table_fly_yeast_human_genegroups"
author: "Vera"
date: "14/02/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

## Objectives:
Create a table containing all genegroup-wise median correlation values for internal and external correlations, as well as p-value and FDR of the shift, based on the datasets of fly genetic interaction profile correlation, fly morphology feature correlation, yeast genetic interaction profile similarity values (pearson correlation) and human gene effect correlations and generate visualizations (Venn diagrams, jitter plots).

## Chunk Description:
- Load packages and data (chunks 1 & 2)
- Familiarize with genegroup dataset (chunk 3) and add genegroup annotation to orthologues list (chunk 4)
- Tidy datasets to keep only genes which have genegroup annotations, and with n >= 4 gene pairs per group (chunks 5-7)
- Calculate p-values (and FDRs) of internal vs. external shift (chunks 8-10)
- Calculate median internal/external correlations per genegroup for each dataset (chunks 11-13)
- Create one dataframe containing internal median, external median, pvalue/FDR for each organism (chunks 14-16)
- Create common dataframe for all dataset (long-format and wide-format) (chunk 17)
- Draw Venn diagrams showing overlaps of significant (FDR < 0.2) internal vs. external shifts (chunk 18)
- Which genegroups show significant (FDR < 0.2) shifts in all datasets? (chunk 19)
- Jitter plots to visualize median of internal correlation vs. median of external correlations (chunk 20)


## 1. Load packages

```{r}

library(tidyverse)
library(dplyr)
library(ggplot2)
library(VennDiagram)

```

## 2. Load datasets

```{r}

# Annotation datasets:

# Load gene group annotations (fly), (Florian, 14.02.2020):
genegroups <- read_delim(file = "internal_vs_external_table/genes_to_gene_groups.txt", delim = "\t")

# Load list of orthologue pairs from Ensembl database -27.11.2019:
orthologues <- read_delim("conserved_functional_groups/new_orthologues_list_copy.txt", delim = "\t")

# Correlation datasets:

# Genetic interaction profile correlations for Fly (based on cell count), (Florian, 29.01.2020):
load(file = "conserved_functional_groups/cell_count_interaction_correlations_pearson_p_unfiltered_fly.RData")
# This dataset is named GI_corrs, in the markdown mostly referred to as "cellcount"

# Morphology feature correlations for Fly, (Florian, 03.02.2020):
load(file = "conserved_functional_groups/feature_correlations_pearson_p_unfiltered_fly.RData")
# This dataset is named feature_corrs

# Genetic interaction profile similarity values for Yeast:
# Read genetic interaction profile similarity values (yeast, https://thecellmap.org/costanzo2016/ - downloaded Jan 27, 2020 (Data File S3))):
# Description: Matrix file containing genetic interaction profile similarity values (as measured by Pearson correlation) for every pair of mutant strains in the dataset. Similarity values were computed for all gene pairs combined (ALL). The matrix contains 2 sets of row and column headers, providing a unique allele name for every mutant strain (row & column header #1) as well as a systematic ORF name (row & column header #2).
yeast_correlation <- read_delim("source_data/costanzo_2016/Data File S3. Genetic interaction profile similarity matrices/cc_ALL.txt", delim = "\t")

# Load human dataset (gene effect correlation)
# Load dataset filtered for genes in the orthologues list (created in chunk 4 of cons_functional_groups_human markdown)
# (otherwise, I'll have the problem of having 168 million rows again)
ceres_corrs_sep_in_ortho <- read_csv(file = "conserved_functional_groups/pre-data/Human_dataset_filtering/ceres_corrs_sep_filtered.csv")

```

## 3. Familiarize with and tidy up new genegroups-dataset

```{r}

dim(genegroups)
# [1] 8214    3

# This is what the dataset looks like:
head(genegroups)
# A tibble: 6 x 3
#  gene_group current_symbol current_fbgn
#  <chr>      <chr>          <chr>       
#1 MYOSINS    jar            FBgn0011225 
#2 MYOSINS    didum          FBgn0261397 
#3 MYOSINS    ninaC          FBgn0002938 
#4 MYOSINS    Myo31DF        FBgn0086347 
#5 MYOSINS    Myo81F         FBgn0267431 
#6 MYOSINS    d              FBgn0262029 

dim(genegroups %>% drop_na() %>% distinct())
# [1] 7742    3       # NAs because some gene groups are written in there, but do not have any genes
# NA-rows are not of any use and are dropped; duplicated rows are dropped as well
genegroups <- genegroups %>% drop_na() %>% distinct()

# How many different gene groups are there?
nrow(genegroups %>% select(gene_group) %>% distinct())
# [1] 808 gene groups

# Are symbols and FBgns unique?
nrow(genegroups)
# [1] 7742
nrow(genegroups %>% select(current_symbol) %>% distinct())
# [1] 6286 -> some genes (gene symbols) have multiple gene group annotations
nrow(genegroups %>% select(current_fbgn) %>% distinct())
# [1] 6306 -> some genes (FBgns) have multiple gene group annotations
nrow(genegroups %>% select(-gene_group) %>% distinct())
# [1] 6306
# There are more rows in total than symbols and FBgns -> some symbols/FBgns are annotated with more than one gene group
# Also, some gene symbols are annotated wth more than one FBgn

# How many genes have more than one genegroup annotation?
nrow(genegroups %>% 
       select(-current_fbgn) %>% 
       distinct() %>% 
       group_by(current_symbol) %>% 
       filter(n() >= 2) %>% 
       select(current_symbol) %>% 
       distinct())
# [1] 982
# Almost 1000 genes have more than one gene group annotation

nrow(genegroups %>% 
       select(-current_fbgn) %>% 
       distinct() %>% 
       group_by(current_symbol) %>% 
       filter(n() >= 12) %>% 
       select(current_symbol) %>% 
       distinct())
# [1] 5 
# 5 genes have >= 12 genegroup annotations

# Also, there are more FBgns than symbols -> some symbols are annotated with more than one FBgn
# Find out, which genes have more than one FBgn and if they are annotated with the same gene group
# which genes have more than one FBgn?
(genegroups %>% 
  select(current_symbol, current_fbgn) %>% 
  distinct() %>% 
  group_by(current_symbol) %>% 
  filter(n () >= 2) %>%
  select(current_symbol) %>% 
  distinct())$current_symbol
# [1] "hoip"     "caz"      "hll"      "gry"      "Pdi"      "mRpS34"   "para"     "cv-c"     "scb"      "SRPK"     "RpS12"    "RpS10b"   "AdamTS-A" "eEF5"     "RpL19"   
# [16] "RpL27A"   "RpL31"    "Rab1"     "Orc1"     "TER94"    
# Filter for those genes, see, if they are annotated with more than one gene group
genegroups %>% 
  filter(current_symbol %in% (genegroups %>% 
                                select(current_symbol, current_fbgn) %>% 
                                distinct() %>% 
                                group_by(current_symbol) %>% 
                                filter(n () >= 2) %>%
                                select(current_symbol) %>% 
                                distinct())$current_symbol) %>% 
  view()
# It seems, that the genes with more than one FBgn have the same group annotation for both FBgns
# Verify:
(genegroups %>% 
  filter(current_symbol %in% (genegroups %>% 
                                select(current_symbol, current_fbgn) %>% 
                                distinct() %>% 
                                group_by(current_symbol) %>% 
                                filter(n () >= 2) %>%
                                select(current_symbol) %>% 
                                distinct())$current_symbol) %>% 
  group_by(gene_group, current_symbol) %>% 
  filter(n() != 2))$current_symbol
# character(0)

# Just to make sure that nothing gets lost when adding the gene group annotation to the orthologues list, I'll use the FBgns for joining (chunk 4)

```

## 4. Add gene group annotation to orthologues list, tidy up resulting dataframe

```{r}

# we wanna keep only those rows of the orthologues list, which contain FBgns with gene group annotation -> use inner_join
orthologues_genegroups <- inner_join(orthologues, (genegroups %>% rename(Gene_stable_ID = current_fbgn)), by = "Gene_stable_ID")
dim(orthologues_genegroups)
# [1] 31386    12

# Check, if Gene_name (from orthologues list) and current_symbol (from gene groups list) are the same (they should be)
which(! identical(orthologues_genegroups$Gene_name, orthologues_genegroups$current_symbol))
# [1] 1         # Apparently, the first entry in the columns is different. 
# However:
identical(orthologues_genegroups$Gene_name[1], orthologues_genegroups$current_symbol[1])
# [1] TRUE
orthologues_genegroups$Gene_name[1]
# [1] "SkpA"
orthologues_genegroups$current_symbol[1]
# [1] "SkpA"

# Convert S_cerevisiae_gene_name to lowercase letters (because they are lowercase in the yeast correlation dataset)
orthologues_genegroups <- orthologues_genegroups %>% 
  mutate(S_cerevisiae_gene_name = tolower(S_cerevisiae_gene_name))

```

## 5.-7. Tidy up datasets

The following chunks are tidying the datasets. Goals: keep only rows, for which there is a genegroup annotation in orthologues_genegroups; add genegroup-annotation for both genes and generate column "type" describing if the measured correlation refers to internal or external genes; filter datasets, to keep only gene groups, for which internal and external correlations have a group size of n >= 4.
Note: after adding genegroup annotations for both genes, the dataframes get longer, due the fact that there are multiple genegroup annotations per gene (at least for some genes).

## 5. Tidy up fly datasets

```{r}

# Use Florian's code to separate gene and IDs into single columns:
# For cell count-based data:
fly_correlation_cellcount <- GI_corrs %>%
  separate(x,c("genex","idx"),sep = "öö")%>%
  separate(y,c("geney","idy"),sep = "öö") %>%
  dplyr::select(genex,geney,idx,idy,r) 
dim(fly_correlation_cellcount)
# [1] 23526370        5
# For feature-based data:
fly_correlation_features <- feature_corrs %>%
  separate(x,c("genex","idx"),sep = "öö")%>%
  separate(y,c("geney","idy"),sep = "öö") %>%
  dplyr::select(genex,geney,idx,idy,r) 
dim(fly_correlation_features)
# [1] 23526370        5

# Filter datasets for genes present in orthologue_genegroups list
fly_corr_cellcount_in_ortho <- fly_correlation_cellcount %>% 
  filter(genex %in% orthologues_genegroups$Gene_name & geney %in% orthologues_genegroups$Gene_name)
dim(fly_corr_cellcount_in_ortho)
# [1] 4840716       5
fly_corr_features_in_ortho <- fly_correlation_features %>% 
  filter(genex %in% orthologues_genegroups$Gene_name & geney %in% orthologues_genegroups$Gene_name)
dim(fly_corr_features_in_ortho)
# [1] 4840716       5

# Create new columns in both datasets (genegroup_1 and genegroup_2), by left-joining with the orthologues_genegroups list and add column "type"
# Cell count-based data:
fly_corr_cellcount_anno <- fly_corr_cellcount_in_ortho %>% 
  rename(Gene_name = genex) %>% 
  left_join((orthologues_genegroups %>% select(Gene_name, gene_group) %>% drop_na() %>% distinct()), by = "Gene_name") %>%
  rename(Gene_1 = Gene_name, genegroup_1 = gene_group) %>% 
  rename(Gene_name = geney) %>% 
  left_join((orthologues_genegroups %>% select(Gene_name, gene_group) %>% drop_na() %>% distinct()), by = "Gene_name") %>% 
  rename(Gene_2 = Gene_name, genegroup_2 = gene_group) %>% 
  mutate(type = ifelse(genegroup_1 == genegroup_2, "internal", "external"))
dim(fly_corr_cellcount_anno)
# [1] 7793867       8
# More rows than before, probably due to the fact, that there are genes annotated with more than one gene group

# Feature-based data:
fly_corr_features_anno <- fly_corr_features_in_ortho %>% 
  rename(Gene_name = genex) %>% 
  left_join((orthologues_genegroups %>% select(Gene_name, gene_group) %>% drop_na() %>% distinct()), by = "Gene_name") %>%
  rename(Gene_1 = Gene_name, genegroup_1 = gene_group) %>% 
  rename(Gene_name = geney) %>% 
  left_join((orthologues_genegroups %>% select(Gene_name, gene_group) %>% drop_na() %>% distinct()), by = "Gene_name") %>% 
  rename(Gene_2 = Gene_name, genegroup_2 = gene_group) %>% 
  mutate(type = ifelse(genegroup_1 == genegroup_2, "internal", "external"))
dim(fly_corr_features_anno)
# [1] 7793867       8
# More rows than before, probably due to the fact, that there are genes annotated with more than one gene group

#### n >= 4, correlation both internal and external ####

# When calculating pvalues (chunk 8), we found out that there are some gene groups that do have more than 4 gene members, but might have only one gene pair in the internal correlations. Because of the division through (n-1) during pvalue-calculation, those rows are a problem, because if n = 1, this would result in dividing through 0.

# -> The datasets are supposed to have n >= 4 gene pairs for both internal and external correlations
# we have to create a vector of gene groups which do not have n >= 4 gene pairs in internal/external correlations and filter those groups out completely (otherwise, there might be gene pairs for external, but not for internal, which also results in problems with p-value calculation)

# Create a vector of gene groups with less than 4 gene pairs for internal corelations
fly_not_n4_int <- (fly_corr_cellcount_anno %>% 
  filter(type == "internal") %>% 
  group_by(genegroup_1) %>% 
  filter(n() < 4) %>% 
  select(genegroup_1) %>% 
  distinct())$genegroup_1

# For external, this is no problem; there are more than 4 gene pairs, no matter which genegroup is the grouping variable:
fly_corr_cellcount_anno %>% 
  filter(type == "external") %>% 
  group_by(genegroup_1) %>% 
  filter(n() < 4) %>% 
  select(genegroup_1) %>% 
  distinct()
# A tibble: 0 x 1
# Groups:   genegroup_1 [0]
# … with 1 variable: genegroup_1 <chr>
fly_corr_cellcount_anno %>% 
  filter(type == "external") %>% 
  group_by(genegroup_2) %>% 
  filter(n() < 4) %>% 
  select(genegroup_2) %>% 
  distinct()
# A tibble: 0 x 1
# Groups:   genegroup_2 [0]
# … with 1 variable: genegroup_2 <chr>

# Filter datasets for gene groups which are not part of the fly_not_n4_int vector:
fly_corr_cellcount_n4 <- fly_corr_cellcount_anno %>% 
  filter(! (genegroup_1 %in% fly_not_n4_int | genegroup_2 %in% fly_not_n4_int))
dim(fly_corr_cellcount_n4)
# [1] 5783906       8
fly_corr_features_n4 <- fly_corr_features_anno %>% 
  filter(! (genegroup_1 %in% fly_not_n4_int | genegroup_2 %in% fly_not_n4_int))
dim(fly_corr_features_n4)
# [1] 5783906       8

# Next problem: there are some gene groups which only have external correlations:
fly_corr_cellcount_n4 %>% select(genegroup_2, type) %>% distinct() %>% group_by(genegroup_2) %>% filter(n () < 2) %>% view()

# Create genegroup-filtering-vector based only on internal correlations to avoid this problem:
genegroups_fly_int <- (fly_corr_cellcount_n4 %>%  filter(type == "internal") %>%  select(genegroup_1) %>% distinct())$genegroup_1 %>% sort()

# Filter datasets for gene groups in the genegroups_fly_int vector:
fly_corr_cellcount_n4 <- fly_corr_cellcount_n4 %>% 
  filter(genegroup_1 %in% genegroups_fly_int, genegroup_2 %in% genegroups_fly_int)
dim(fly_corr_cellcount_n4)
# [1] 5191114       8
fly_corr_features_n4 <- fly_corr_features_n4 %>% 
  filter(genegroup_1 %in% genegroups_fly_int, genegroup_2 %in% genegroups_fly_int)
dim(fly_corr_features_n4)
# [1] 5191114       8

```

## 6. Tidy up yeast dataset

```{r}

# Drop ORF columns
dim(yeast_correlation)
# [1] 6231 6232
yeast_correlation <- yeast_correlation %>% 
  rename(allele_name = X1) %>%    # rename column for allele
  select(-X2) %>%                 # drop column X2 (= column containing ORF names)
  drop_na()                       # there is only one row containing NAs which is the first row, which has the ORF names, so I'll just use the drop_na() function to get rid of the row
dim(yeast_correlation)
# 6230 6231 <- one column and one row less than before 

# There are duplicates (due to the fact that the matrix is symmetrical), for example, there is gene_name_1 aap1 and gene_name_2 aat2, as well as gene_name_1 aat2 and gene_name_2 aap1
# example: 
# aat2  NA  aap1  NA  -0.05033
# aap1  NA  aat2  NA  -0.05033
# therefore: set upper triangle to be NA

# Set column allele_names as rownames
yeast_correlation <- yeast_correlation %>% 
  column_to_rownames(var = "allele_name")
# first, we need to check whether the column names and row names are in the same order:
identical(rownames(yeast_correlation), colnames(yeast_correlation))
# [1] TRUE        # fortunately, they are

# Replace upper triangle of the matrix with NA, including the diagonale
yeast_correlation[upper.tri(yeast_correlation, diag=TRUE)] <- NA 

# Convert rownames back to column
yeast_correlation <- yeast_correlation %>% 
  rownames_to_column(var = "allele_name")

# Gather to get long-format
yeast_correlation_gathered <- yeast_correlation %>% 
  gather(allele_name_2, value, -allele_name)
dim(yeast_correlation_gathered)
#[1] 38812900        3

# Drop the NAs. Additionally, for some allele combinations, the value is "NaN"" - we cannot use those, so I'll filter them out
yeast_correlation_gathered <- yeast_correlation_gathered %>% 
  drop_na() %>% 
  filter(value != "NaN", value != "NA")     # for whatever reason, the drop_na() did not work, but filtering for the string "NA" did
dim(yeast_correlation_gathered)
# [1] 17516577        3 
# Less than half of the rows. Makes sense, because diagonal is excluded

# Separate allele_name into gene name and allele name:
# Using merge (allows only as many separations as number of new columns we give):
yeast_correlation_gathered_sep <- yeast_correlation_gathered %>% 
  separate(allele_name, into = c("gene_name_1", "allele_name_1"), sep = "-", extra = "merge") %>% 
  separate(allele_name_2, into = c("gene_name_2", "allele_name_2"), sep = "-", extra = "merge")

# Filter yeast dataset for genes present in the orthologues_genegroups list
yeast_corr_in_ortho <- yeast_correlation_gathered_sep %>% 
  filter(gene_name_1 %in% orthologues_genegroups$S_cerevisiae_gene_name & gene_name_2 %in% orthologues_genegroups$S_cerevisiae_gene_name)
dim(yeast_corr_in_ortho)
# [1] 339475      5

# Create new columns in this dataset (genegroup_1 and genegroup_2), by left-joining with the orthologues_genegroups list and add column "type"
yeast_corr_anno <- yeast_corr_in_ortho %>% 
  rename(S_cerevisiae_gene_name = gene_name_1) %>% 
  left_join((orthologues_genegroups %>% select(S_cerevisiae_gene_name, gene_group) %>% drop_na() %>% distinct()), by = "S_cerevisiae_gene_name") %>%
  rename(gene_name_1 = S_cerevisiae_gene_name, genegroup_1 = gene_group) %>% 
  rename(S_cerevisiae_gene_name = gene_name_2) %>% 
  left_join((orthologues_genegroups %>% select(S_cerevisiae_gene_name, gene_group) %>% drop_na() %>% distinct()), by = "S_cerevisiae_gene_name") %>%
  rename(gene_name_2 = S_cerevisiae_gene_name, genegroup_2 = gene_group) %>% 
  mutate(type = ifelse(genegroup_1 == genegroup_2, "internal", "external"))
dim(yeast_corr_anno)
# [1] 853329      8

#### n >= 4, correlation both internal and external ####

# Create a vector of gene groups with less than 4 gene pairs for internal correlations
yeast_not_n4_int <- (yeast_corr_anno %>% 
  filter(type == "internal") %>% 
  group_by(genegroup_1) %>% 
  filter(n() < 4) %>% 
  select(genegroup_1) %>% 
  distinct())$genegroup_1

# For external, this is no problem; there are more than 4 gene pairs, no matter which genegroup is the grouping variable:
yeast_corr_anno %>% 
  filter(type == "external") %>% 
  group_by(genegroup_1) %>% 
  filter(n() < 4) %>% 
  select(genegroup_1) %>% 
  distinct()
# A tibble: 0 x 1
# Groups:   genegroup_1 [0]
# … with 1 variable: genegroup_1 <chr>
yeast_corr_anno %>% 
  filter(type == "external") %>% 
  group_by(genegroup_2) %>% 
  filter(n() < 4) %>% 
  select(genegroup_2) %>% 
  distinct()
# A tibble: 0 x 1
# Groups:   genegroup_2 [0]
# … with 1 variable: genegroup_2 <chr>

# Filter datasets for gene groups which are not part of the yeast_not_n4_int vector:
yeast_corr_n4 <- yeast_corr_anno %>% 
  filter(! (genegroup_1 %in% yeast_not_n4_int | genegroup_2 %in% yeast_not_n4_int))
dim(yeast_corr_n4)
# [1] 525649      8

# Next problem: there are some gene groups which only have external correlations:
yeast_corr_n4 %>% select(genegroup_2, type) %>% distinct() %>% group_by(genegroup_2) %>% filter(n () < 2) %>% view()

# Create genegroup-filtering-vector based only on internal correlations to avoid this problem:
genegroups_yeast_int <- (yeast_corr_n4 %>%  filter(type == "internal") %>%  select(genegroup_1) %>% distinct())$genegroup_1 %>% sort()

# Filter dataset for gene groups in the genegroups_yeast_int vector:
yeast_corr_n4 <- yeast_corr_n4 %>% 
  filter(genegroup_1 %in% genegroups_yeast_int, genegroup_2 %in% genegroups_yeast_int)
dim(yeast_corr_n4)
# [1] 420369      8

```

## 7. Tidy up human dataset

```{r}

# The human dataset was already filtered to contain only genes present in the unmodified orthologues list (see cons_functional_groups_human markdown)
# But: needs to be filtered for genes with gene group annotation ( = genes which are in orthologues_genegroups)

dim(ceres_corrs_sep_in_ortho)
# [1] 37208251        5

# Filter for genes present in orthologues_genegroups list:
ceres_corrs_sep_in_ortho_anno <- ceres_corrs_sep_in_ortho %>% 
  select(-xnum, -ynum) %>% 
  filter(x %in% orthologues_genegroups$Human_gene_name & y %in% orthologues_genegroups$Human_gene_name)
dim(ceres_corrs_sep_in_ortho_anno)
# [1] 11613790        3

# Create new columns in this dataset (genegroup_1 and genegroup_2), by left-joining with the orthologues_genegroups list and add column "type"
human_corr_anno <- ceres_corrs_sep_in_ortho_anno %>%
  rename(Human_gene_name = x) %>%                                                                                                       # rename to enable join by
  left_join((orthologues_genegroups %>% select(Human_gene_name, gene_group) %>% drop_na() %>% distinct()), by = "Human_gene_name") %>%  # join genegroup annotation first gene
  rename(Gene_1 = Human_gene_name, genegroup_1 = gene_group) %>%                                                                        # rename to avoid duplications 
  rename(Human_gene_name = y) %>%                                                                                                       # rename to enable join by
  left_join((orthologues_genegroups %>% select(Human_gene_name, gene_group) %>% drop_na() %>% distinct()), by = "Human_gene_name") %>%  # join complex annotations second gene
  rename(Gene_2 = Human_gene_name, genegroup_2 = gene_group) %>%                                                                        # rename to keep naming consistent
  mutate(type = ifelse(genegroup_1 == genegroup_2, "internal", "external"))                                         # add column "type" to specify internal vs. external
dim(human_corr_anno)
# [1] 23515979        6

#### n >= 4, correlation both internal and external ####

# Create a vector of gene groups with less than 4 gene pairs for internal correlations
human_not_n4_int <- (human_corr_anno %>% 
  filter(type == "internal") %>% 
  group_by(genegroup_1) %>% 
  filter(n() < 4) %>% 
  select(genegroup_1) %>% 
  distinct())$genegroup_1

# For external, this is no problem; there are more than 4 gene pairs, no matter which genegroup is the grouping variable:
human_corr_anno %>% 
  filter(type == "external") %>% 
  group_by(genegroup_1) %>% 
  filter(n() < 4) %>% 
  select(genegroup_1) %>% 
  distinct()
# A tibble: 0 x 1
# Groups:   genegroup_1 [0]
# … with 1 variable: genegroup_1 <chr>
human_corr_anno %>% 
  filter(type == "external") %>% 
  group_by(genegroup_2) %>% 
  filter(n() < 4) %>% 
  select(genegroup_2) %>% 
  distinct()
# A tibble: 0 x 1
# Groups:   genegroup_2 [0]
# … with 1 variable: genegroup_2 <chr>

# Filter datasets for gene groups which are not part of the human_not_n4_int vector:
human_corr_n4 <- human_corr_anno %>% 
  filter(! (genegroup_1 %in% human_not_n4_int | genegroup_2 %in% human_not_n4_int))
dim(human_corr_n4)
# [1] 21242037        6

# Next problem: there are some gene groups which only have external correlations:
human_corr_n4 %>% select(genegroup_1, type) %>% distinct() %>% group_by(genegroup_1) %>% filter(n () < 2) %>% view()

# Create genegroup-filtering-vector based only on internal correlations to avoid this problem:
genegroups_human_int <- (human_corr_n4 %>%  filter(type == "internal") %>%  select(genegroup_1) %>% distinct())$genegroup_1 %>% sort()

# Filter dataset for gene groups in the genegroups_yeast_int vector:
human_corr_n4 <- human_corr_n4 %>% 
  filter(genegroup_1 %in% genegroups_human_int, genegroup_2 %in% genegroups_human_int)
dim(human_corr_n4)
# [1] 20723776        6

```

## 8.-10. Calculate p-values and FDRs of internal vs. external shift

In chunks 8 - 10 , p-values of shifts between internal and external correlations are calculated per gene group and FDR correction is performed for each dataset individually 

## 8. Calculate p-values of internal vs. external shift, FDR correction - fly

```{r}

# Calculate p-values, create common dataframe with p-values for both fly datasets:
# Create new empty dataframe
tt_fly <- data.frame(matrix(ncol = 3, nrow = length(genegroups_fly_int))) %>% 
  rename(genegroup = X1, pval_cellcount = X2, pval_features = X3)

# Calculate p-values, internal vs. external, for cellcount- and feature-based data
j <- 1
for (i in (genegroups_fly_int)) {
    tt_fly$genegroup[j] <- as.character(i)
    tt_fly$pval_cellcount[j] <- t.test(formula = r ~type, alternative = c("two.sided"), data = fly_corr_cellcount_n4 %>%  filter(genegroup_1 == we | genegroup_2 == i))$p.value
    tt_fly$pval_features[j] <- t.test(formula = r ~type, alternative = c("two.sided"), data = fly_corr_features_n4 %>%  filter(genegroup_1 == we | genegroup_2 == i))$p.value
    j <- j+1
}  

# Add columns with FDR-corrected pvalues:
tt_fly <- tt_fly %>% 
  mutate(fdr_cellcount = p.adjust(pval_cellcount, method = "BH")) %>% 
  mutate(fdr_features = p.adjust(pval_features, method = "BH"))

```

## 9. Calculate p-values of internal vs. external shift, FDR correction - yeast

```{r}

# Calculate p-values for yeast dataset:
# Create new empty dataframe
tt_yeast<- data.frame(matrix(ncol = 2, nrow = length(genegroups_yeast_int))) %>% 
  rename(genegroup = X1, pval_cellcount = X2)

# Calculate p-values, internal vs. external
j <- 1
for (i in (genegroups_yeast_int)) {
    tt_yeast$genegroup[j] <- as.character(i)
    tt_yeast$pval_cellcount[j] <- t.test(formula = as.numeric(value) ~type, alternative = c("two.sided"), data = yeast_corr_n4 %>%  filter(genegroup_1 == we | genegroup_2 == i))$p.value
    j <- j+1
}  

# Add columns with FDR-corrected pvalues:
tt_yeast <- tt_yeast %>% 
  mutate(fdr_cellcount = p.adjust(pval_cellcount, method = "BH"))

```

## 10. Calculate p-values of internal vs. external shift, FDR correction - human

```{r}

# Calculate p-values for human dataset:
# Create new empty dataframe
tt_human<- data.frame(matrix(ncol = 2, nrow = length(genegroups_human_int))) %>% 
  rename(genegroup = X1, pval_cellcount = X2)

# Calculate p-values, internal vs. external
j <- 1
for (i in (genegroups_human_int)) {
    tt_human$genegroup[j] <- as.character(i)
    tt_human$pval_cellcount[j] <- t.test(formula = r ~type, alternative = c("two.sided"), data = human_corr_n4 %>%  filter(genegroup_1 == we | genegroup_2 == i))$p.value
    j <- j+1
}  

# Add columns with FDR-corrected pvalues:
tt_human<- tt_human %>% 
  mutate(fdr_cellcount = p.adjust(pval_cellcount, method = "BH"))

```

## 11.-13. Calculate median internal/external correlation values

In chunks 11 - 13, datasets are separated into internal and external in order to calculate median correlation values based on gene group and correlation type

## 11. Calculate median correlation - fly

```{r}

#### For genetic interaction profile correlation ####

# Median correlation values for internal correlations, based on gen. int. profile (cellcount)
fly_corr_cellcount_int <- fly_corr_cellcount_n4 %>% 
  filter(type == "internal") %>%                          # we wanna calculate the median r value for internal correlations
  select(genegroup_1, r) %>%                              # the other columns are not needed anymore, genegroup_1 = genegroup_2 anyway (because of filtering for internal)
  group_by(genegroup_1) %>%                               # calculation of median r value should be done per genegroup
  mutate(median_r_GI_int = median(r)) %>%                 # calculate median r value
  select(-r) %>%                                          # r values are not needed anymore
  distinct() %>%                                          # remove duplicated rows to get one row per genegroup, containing the genegroup and the median r value
  rename(genegroup = genegroup_1)

# Median correlation values for external correlations, based on genetic interaction profile correlation
# A little more complicated, because in this case, genegroup_1 != genegroup_2 -> for-loop is needed to filter genegroup columns
# Create new empty dataframe
fly_corr_cellcount_ext = as_tibble(matrix(nrow = length(genegroups_fly_int), ncol = 2)) %>% 
  rename(genegroup = V1, median_r_GI_ext = V2)

# Calculate median of external correlation
j <- 1
for (i in genegroups_fly_int) {
  fly_corr_cellcount_ext$genegroup[j] <- as.character(i)                                                      # at position j, write name of complex
  fly_corr_cellcount_ext$median_r_GI_ext[j] <- (fly_corr_cellcount_n4 %>% 
                                                            filter(type == "external") %>%                    # we wanna have median r values for external correlations
                                                            filter(genegroup_1 == we | genegroup_2 == i) %>%   # one genegroup has to be in i
                                                            select(-Gene_1, -Gene_2, -idx, -idy) %>%          # those columns are not needed anymore
                                                            mutate (median_r_GI_ext = median(r)) %>%          # calculate median correlation value
                                                            select(median_r_GI_ext) %>%                       # all the other rows are not needed anymore
                                                            distinct()
                                                            )$median_r_GI_ext[1]                              # select first item of column, -> new dataframe (position j)
    j <- j + 1
}

#### For morphology feature correlation ####

# Median correlation values for internal correlations, based on morphology features
fly_corr_features_int <- fly_corr_features_n4 %>% 
  filter(type == "internal") %>%                          # we wanna calculate the median r value for internal correlations
  select(genegroup_1, r) %>%                              # the other columns are not needed anymore, geengroup_1 = genegroup_2 anyway (because of filtering for internal)
  group_by(genegroup_1) %>%                               # calculation of median r value should be done per genegroup
  mutate(median_r_features_int = median(r)) %>%           # calculate median r value
  select(-r) %>%                                          # r values are not needed anymore
  distinct() %>%                                          # remove duplicated rows to get one row per genegoup, containing the genegroup and the median r value
  rename(genegroup = genegroup_1)

# Median correlation values for external correlations, based on morphology feature correlation
# A little more complicated, because in this case, genegroup_1 != genegroup_2 -> for-loop is needed to filter genegroup columns
# Create new empty dataframe
fly_corr_features_ext = as_tibble(matrix(nrow = length(genegroups_fly_int), ncol = 2)) %>% 
  rename(genegroup = V1, median_r_features_ext = V2)

# Calculate median of external correlation
j <- 1
for (i in genegroups_fly_int) {
  fly_corr_features_ext$genegroup[j] <- as.character(i)                                                       # at position j, write name of complex
  fly_corr_features_ext$median_r_features_ext[j] <- (fly_corr_features_n4 %>% 
                                                            filter(type == "external") %>%                    # we wanna have median r values for external correlations
                                                            filter(genegroup_1 == we | genegroup_2 == i) %>%   # one genegroup has to be in i
                                                            select(-Gene_1, -Gene_2, -idx, -idy) %>%          # those columns are not needed anymore
                                                            mutate (median_r_features_ext = median(r)) %>%    # calculate median correlation value
                                                            select(median_r_features_ext) %>%                 # all the other rows are not needed anymore
                                                            distinct()
                                                            )$median_r_features_ext[1]                        # select first item of column, -> new dataframe (position j)
    j <- j + 1
}

```

## 12. Calculate median correlation - yeast

```{r}

# Median correlation values for internal correlations
yeast_corr_int <- yeast_corr_n4 %>% 
  filter(type == "internal") %>%                          # we wanna calculate the median r value for internal correlations
  select(genegroup_1, value) %>%                          # the other columns are not needed anymore, genegroup_1 = genegroup_2 anyway (because of filtering for internal)
  group_by(genegroup_1) %>%                               # calculation of median r value should be done per genegroup
  mutate(median_r_GI_int = median(as.numeric(value))) %>% # calculate median r value
  select(-value) %>%                                      # r values are not needed anymore
  distinct() %>%                                          # remove duplicated rows to get one row per genegroup, containing the genegroup and the median r value
  rename(genegroup = genegroup_1)

# Median correlation values for external correlations
# A little more complicated, because in this case, genegroup_1 != genegroup_2 -> for-loop is needed to filter genegroup columns
# Create new empty dataframe
yeast_corr_ext = as_tibble(matrix(nrow = length(genegroups_yeast_int), ncol = 2)) %>% 
  rename(genegroup = V1, median_r_GI_ext = V2)

# Calculate median of external correlations
j <- 1
for (i in genegroups_yeast_int) {
  yeast_corr_ext$genegroup[j] <- as.character(i)                                                              # at position j, write name of complex
  yeast_corr_ext$median_r_GI_ext[j] <- (yeast_corr_n4 %>% 
                                                            filter(type == "external") %>%                    # we wanna have median r values for external correlations
                                                            filter(genegroup_1 == we | genegroup_2 == i) %>%   # one genegroup has to be in i
                                                            select(-gene_name_1, -gene_name_2, -allele_name_1, -allele_name_2) %>%  # those columns are not needed anymore
                                                            mutate (median_r_GI_ext = median(as.numeric(value))) %>%    # calculate median correlation value
                                                            select(median_r_GI_ext) %>%                       # all the other rows are not needed anymore
                                                            distinct()
                                                            )$median_r_GI_ext[1]                              # select first item of column, -> new dataframe (position j)
    j <- j + 1
}

```

## 13. Calculate median correlation - human

```{r}

# Median correlation values for internal correlations
human_corr_int <- human_corr_n4 %>% 
  filter(type == "internal") %>%                          # we wanna calculate the median r value for internal correlations
  select(genegroup_1, r) %>%                              # the other columns are not needed anymore, genegroup_1 = genegroup_2 anyway (because of filtering for internal)
  group_by(genegroup_1) %>%                               # calculation of median r value should be done per genegroup
  mutate(median_r_GI_int = median(r)) %>%                 # calculate median r value
  select(-r) %>%                                          # r values are not needed anymore
  distinct() %>%                                          # remove duplicated rows to get one row per genegroup, containing the genegroup and the median r value
  rename(genegroup = genegroup_1)

# Median correlation values for external correlations
# A little more complicated, because in this case, genegroup_1 != genegroup_2 -> for-loop is needed to filter genegroup columns
# Create new empty dataframe
human_corr_ext = as_tibble(matrix(nrow = length(genegroups_human_int), ncol = 2)) %>% 
  rename(genegroup = V1, median_r_GI_ext = V2)

# Calculate median of external correlation
j <- 1
for (i in genegroups_human_int) {
  human_corr_ext$genegroup[j] <- as.character(i)                                                              # at position j, write name of complex
  human_corr_ext$median_r_GI_ext[j] <- (human_corr_n4 %>% 
                                                            filter(type == "external") %>%                    # we wanna have median r values for external correlations
                                                            filter(genegroup_1 == we | genegroup_2 == i) %>%   # one genegroup has to be in i
                                                            select(-Gene_1, -Gene_2) %>%                      # those columns are not needed anymore
                                                            mutate (median_r_GI_ext = median(r)) %>%          # calculate median correlation value
                                                            select(median_r_GI_ext) %>%                       # all the other rows are not needed anymore
                                                            distinct()
                                                            )$median_r_GI_ext[1]                              # select first item of column, -> new dataframe (position j)
    j <- j + 1
}

```

## 14.-16. Create one dataframe containing median internal, median external and p-values for each organism

Dataframes are created by left-joining dataframes for median internal, median external and p values/FDRs

## 14. Dataframe fly

```{r}

# Create dataset containing median internal/external correlation, p-value and FDR for fly (both datasets):
fly_corr_analysis <- left_join(fly_corr_cellcount_int, fly_corr_cellcount_ext, by = "genegroup") %>% 
  left_join((tt_fly %>% select(genegroup, pval_cellcount, fdr_cellcount)), by = "genegroup") %>% 
  left_join(fly_corr_features_int, by = "genegroup") %>% 
  left_join(fly_corr_features_ext, by = "genegroup") %>% 
  left_join((tt_fly %>% select(genegroup, pval_features, fdr_features))) %>% 
  rename(pval_GI = pval_cellcount, fdr_GI = fdr_cellcount)
dim(fly_corr_analysis)
# [1] 301   9

# write_delim(fly_corr_analysis, path = "internal_vs_external_table/internal_vs_external_fly_genintprofile_features_genegroupsortho.txt", delim = "\t")

```

## 15. Dataframe yeast

```{r}

# Create dataset containing median internal/external correlation, p-value and FDR for yeast:
yeast_corr_analysis <- left_join(yeast_corr_int, yeast_corr_ext, by = "genegroup") %>% 
  left_join(tt_yeast, by = "genegroup") %>% 
  rename(pval_GI = pval_cellcount, fdr_GI = fdr_cellcount)
dim(yeast_corr_analysis)
# [1] 98  5

# write_delim(yeast_corr_analysis, path = "internal_vs_external_table/internal_vs_external_yeast_genegroupsortho.txt", delim = "\t")

```

## 16. Dataframe human

```{r}

# Create dataset containing median internal/external correlation, p-value and FDR for human:
human_corr_analysis <- left_join(human_corr_int, human_corr_ext, by = "genegroup") %>% 
  left_join(tt_human, by = "genegroup") %>% 
  rename(pval_GI = pval_cellcount, fdr_GI = fdr_cellcount)
dim(human_corr_analysis)
# [1] 459   5

# write_delim(human_corr_analysis, path = "internal_vs_external_table/internal_vs_external_human_genegroupsortho.txt", delim = "\t")

```

## 17. Common dataframe for all organisms

```{r}

# Create common dataframe (long-format) by rbinding the three sub-dataframes together (and reorder columns)
corr_analysis <- rbind((fly_corr_analysis %>% mutate(organism = "fly")),
                       yeast_corr_analysis %>% mutate(organism = "yeast"),
                       human_corr_analysis %>% mutate(organism = "human")) %>% 
  select(organism, genegroup, median_r_GI_int, median_r_GI_ext, pval_GI, fdr_GI, median_r_features_int, median_r_features_ext, pval_features, fdr_features)

dim(corr_analysis)
# [1] 858  10 - makes sense, because: 301 + 98 + 459 = 858

# write_delim(corr_analysis, path = "internal_vs_external_table/internal_vs_external_fly_yeast_human_genegroupsortho.txt", delim = "\t")

# Create wide-format common dataframe
corr_analysis_wide <- full_join((fly_corr_analysis %>% 
                                   rename(fly_median_r_GI_int = median_r_GI_int, fly_median_r_GI_ext = median_r_GI_ext, fly_pval_GI = pval_GI, fly_fdr_GI = fdr_GI)), 
                                yeast_corr_analysis %>% 
                                  rename(yeast_median_r_GI_int = median_r_GI_int, yeast_median_r_GI_ext = median_r_GI_ext, yeast_pval_GI = pval_GI, yeast_fdr_GI = fdr_GI),
                                by = "genegroup") %>%
  full_join(human_corr_analysis %>% 
              rename(human_median_r_GI_int = median_r_GI_int, human_median_r_GI_ext = median_r_GI_ext, human_pval_GI = pval_GI, human_fdr_GI = fdr_GI),
            by = "genegroup")
dim(corr_analysis_wide)
#[1] 502  17
# The wide-format table contains lots of NAs, in case there is data for a certain genegroup in one organism, but no data for this genegroup in the second organism
nrow(corr_analysis_wide %>% drop_na())
# [1] 87    # there are 87 genegroups for which there is data in every dataset

# write_delim(corr_analysis_wide, path = "internal_vs_external_table/internal_vs_external_fly_yeast_human_genegroupsortho_wide.txt", delim = "\t")

```

## 18. Venn diagrams significance

```{r}

# Venn diagrams illustrating overlaps of significant internal vs. external shifts

#### Pairwise ####

# Fly: significant genetic interaction profile correlations vs significant feature corelations
grid.newpage()
draw.pairwise.venn(area1 = nrow(corr_analysis_wide %>% filter(fly_fdr_GI < 0.2)), 
                   area2 = nrow(corr_analysis_wide %>% filter(fdr_features< 0.2)), 
                   cross.area = nrow(corr_analysis_wide %>%  filter(fly_fdr_GI < 0.2 & fdr_features < 0.2)), 
                   category = c ("gen. int. profile", "morph. features"), fill = c("#4285f4", "#d62d20"), 
                   scaled = FALSE, fontface = 2, lwd = 1, alpha = 0.35, cex = 1.3, cat.cex = 1.3, cat.fontface = 2, cat.pos = c(20,-20), cat.dist = 0.05)
grid.text("Significant internal vs. external shifts, fly", vjust = -17, gp = gpar(fontsize = 15, fontface = "bold"))

#### Triple (datasets human, yeast, one fly dataset each) ####

# Without filtering for genegroups present in all datasets:

# Significant internal vs external shift for all organisms -> overlap (without fly morphology feature correlation)
grid.newpage()
draw.triple.venn(area1 = nrow(corr_analysis_wide %>% filter(fly_fdr_GI < 0.2)),
                 area2 = nrow(corr_analysis_wide %>% filter(yeast_fdr_GI < 0.2)),
                 area3 = nrow(corr_analysis_wide %>% filter(human_fdr_GI < 0.2)),
                 n12 = nrow(corr_analysis_wide %>% filter(fly_fdr_GI < 0.2 & yeast_fdr_GI < 0.2)),
                 n23 = nrow(corr_analysis_wide %>% filter(yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
                 n13 = nrow(corr_analysis_wide %>% filter(fly_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
                 n123 = nrow(corr_analysis_wide %>%  filter(fly_fdr_GI < 0.2 & yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
                 category = c("fly GI", "yeast", "human"),
                 cat.fontface = 2, cat.cex = 1.3, cat.dist = 0.04,
                 fill = c("#4285f4", "#ffa700", "#008744"), alpha = 0.35,
                 fontface = 2, cex = 1.3,
                 lwd = 1
                 )
#dev.copy2pdf(file = "internal_vs_external_table/Plots/fly_yeast_human_genegroups_internalvsexternal_significance_triplevenn_flygi.pdf")

# Significant internal vs. external shift with feature correlations and yeast/human gen. int. correlations
grid.newpage()
draw.triple.venn(area1 = nrow(corr_analysis_wide %>% filter(fdr_features < 0.2)),
                 area2 = nrow(corr_analysis_wide %>% filter(yeast_fdr_GI < 0.2)),
                 area3 = nrow(corr_analysis_wide %>% filter(human_fdr_GI < 0.2)),
                 n12 = nrow(corr_analysis_wide %>% filter(fdr_features < 0.2 & yeast_fdr_GI < 0.2)),
                 n23 = nrow(corr_analysis_wide %>% filter(yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
                 n13 = nrow(corr_analysis_wide %>% filter(fdr_features < 0.2 & human_fdr_GI < 0.2)),
                 n123 = nrow(corr_analysis_wide %>%  filter(fdr_features < 0.2 & yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
                 category = c("fly features", "yeast", "human"),
                 cat.fontface = 2, cat.cex = 1.3, cat.dist = 0.06,
                 fill = c("#d62d20", "#ffa700", "#008744"), alpha = 0.35,
                 fontface = 2, cex = 1.3,
                 lwd = 1
                 )
#dev.copy2pdf(file = "internal_vs_external_table/Plots/fly_yeast_human_genegroups_internalvsexternal_significance_triplevenn_flyfeatures.pdf")

# With filtering for genegroups present in all datasets:

# Significant internal vs external shift for all organisms -> overlap (without fly morphology feature correlation), filtered for genegroups in all datasets
grid.newpage()
draw.triple.venn(area1 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fly_fdr_GI < 0.2)),
                 area2 = nrow(corr_analysis_wide %>% drop_na() %>% filter(yeast_fdr_GI < 0.2)),
                 area3 = nrow(corr_analysis_wide %>% drop_na() %>% filter(human_fdr_GI < 0.2)),
                 n12 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fly_fdr_GI < 0.2 & yeast_fdr_GI < 0.2)),
                 n23 = nrow(corr_analysis_wide %>% drop_na() %>% filter(yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
                 n13 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fly_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
                 n123 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fly_fdr_GI < 0.2 & yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
                 category = c("fly GI", "yeast", "human"),
                 cat.fontface = 2, cat.cex = 1.3, cat.dist = 0.04,
                 fill = c("#4285f4", "#ffa700", "#008744"), alpha = 0.35,
                 fontface = 2, cex = 1.3,
                 lwd = 1
                 )
#dev.copy2pdf(file = "internal_vs_external_table/Plots/fly_yeast_human_genegroups_internalvsexternal_significance_triplevenn_flygi_filtered.pdf")

# Significant internal vs external shift for all organisms -> overlap with feature correlations and yeast/human gen. int. correlations, filtered for genegroups in all datasets
grid.newpage()
draw.triple.venn(area1 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fdr_features < 0.2)),
                 area2 = nrow(corr_analysis_wide %>% drop_na() %>% filter(yeast_fdr_GI < 0.2)),
                 area3 = nrow(corr_analysis_wide %>% drop_na() %>% filter(human_fdr_GI < 0.2)),
                 n12 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fdr_features < 0.2 & yeast_fdr_GI < 0.2)),
                 n23 = nrow(corr_analysis_wide %>% drop_na() %>% filter(yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
                 n13 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fdr_features < 0.2 & human_fdr_GI < 0.2)),
                 n123 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fdr_features < 0.2 & yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
                 category = c("fly features", "yeast", "human"),
                 cat.fontface = 2, cat.cex = 1.3, cat.dist = 0.06,
                 fill = c("#d62d20", "#ffa700", "#008744"), alpha = 0.35,
                 fontface = 2, cex = 1.3,
                 lwd = 1
                 )
#dev.copy2pdf(file = "internal_vs_external_table/Plots/fly_yeast_human_genegroups_internalvsexternal_significance_triplevenn_flyfeatures_filtered.pdf")

#### Quad (all datasets) ####

# Significant internal vs external shift for all organisms -> overlap (including fly morphology feature correlation)
grid.newpage()
draw.quad.venn(area1 = nrow(corr_analysis_wide %>% filter(fly_fdr_GI < 0.2)),
               area2 = nrow(corr_analysis_wide %>% filter(fdr_features < 0.2)),
               area3 = nrow(corr_analysis_wide %>% filter(yeast_fdr_GI < 0.2)),
               area4 = nrow(corr_analysis_wide %>% filter(human_fdr_GI < 0.2)),
               n12 = nrow(corr_analysis_wide %>% filter(fly_fdr_GI < 0.2 & fdr_features < 0.2)),
               n13 = nrow(corr_analysis_wide %>% filter(fly_fdr_GI < 0.2 & yeast_fdr_GI < 0.2)),
               n14 = nrow(corr_analysis_wide %>% filter(fly_fdr_GI < 0.2, human_fdr_GI < 0.2)),
               n23 = nrow(corr_analysis_wide %>% filter(fdr_features < 0.2 & yeast_fdr_GI < 0.2)),
               n24 = nrow(corr_analysis_wide %>% filter(fdr_features < 0.2 & human_fdr_GI < 0.2)),
               n34 = nrow(corr_analysis_wide %>% filter(yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
               n123 = nrow(corr_analysis_wide %>% filter(fly_fdr_GI < 0.2 & fdr_features < 0.2 & yeast_fdr_GI < 0.2)),
               n124 = nrow(corr_analysis_wide %>% filter(fly_fdr_GI < 0.2 & fdr_features < 0.2 & human_fdr_GI < 0.2)),
               n134 = nrow(corr_analysis_wide %>% filter(fly_fdr_GI < 0.2 & yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
               n234 = nrow(corr_analysis_wide %>% filter(fdr_features < 0.2 & yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
               n1234 = nrow(corr_analysis_wide %>% filter(fly_fdr_GI < 0.2 & fdr_features < 0.2 & yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
               category = c("fly GI", "fly m. f.", "yeast", "human"),
               cat.fontface = 2, cat.cex = 1.3, cat.dist = c(0.2, 0.2, 0.1, 0.1),
               fill = c("#4285f4", "#d62d20", "#ffa700", "#008744"), alpha = 0.35,
               fontface = 2, cex = 1.3,
               lwd = 1
               )
grid.text("Significant internal vs. external shifts, all datasets", vjust = -24, gp = gpar(fontsize = 15, fontface = "bold"))
#dev.copy2pdf(file = "internal_vs_external_table/Plots/fly_yeast_human_genegroups_internalvsexternal_significance_quadvenn.pdf")

# Significant internal vs external shift for all organisms -> overlap (including fly morphology feature correlation), filtered for genegroups in all datasets
grid.newpage()
draw.quad.venn(area1 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fly_fdr_GI < 0.2)),
               area2 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fdr_features < 0.2)),
               area3 = nrow(corr_analysis_wide %>% drop_na() %>% filter(yeast_fdr_GI < 0.2)),
               area4 = nrow(corr_analysis_wide %>% drop_na() %>% filter(human_fdr_GI < 0.2)),
               n12 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fly_fdr_GI < 0.2 & fdr_features < 0.2)),
               n13 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fly_fdr_GI < 0.2 & yeast_fdr_GI < 0.2)),
               n14 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fly_fdr_GI < 0.2, human_fdr_GI < 0.2)),
               n23 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fdr_features < 0.2 & yeast_fdr_GI < 0.2)),
               n24 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fdr_features < 0.2 & human_fdr_GI < 0.2)),
               n34 = nrow(corr_analysis_wide %>% drop_na() %>% filter(yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
               n123 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fly_fdr_GI < 0.2 & fdr_features < 0.2 & yeast_fdr_GI < 0.2)),
               n124 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fly_fdr_GI < 0.2 & fdr_features < 0.2 & human_fdr_GI < 0.2)),
               n134 = nrow(corr_analysis_wide %>% drop_na() %>%  filter(fly_fdr_GI < 0.2 & yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
               n234 = nrow(corr_analysis_wide %>% drop_na() %>% filter(fdr_features < 0.2 & yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
               n1234 = nrow(corr_analysis_wide %>% drop_na() %>%  filter(fly_fdr_GI < 0.2 & fdr_features < 0.2 & yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2)),
               category = c("fly GI", "fly m. f.", "yeast", "human"),
               cat.fontface = 2, cat.cex = 1.3, cat.dist = c(0.2, 0.2, 0.1, 0.1),
               fill = c("#4285f4", "#d62d20", "#ffa700", "#008744"), alpha = 0.35,
               fontface = 2, cex = 1.3,
               lwd = 1
               )
grid.text("Significant internal vs. external shifts, filtered datasets", vjust = -24, gp = gpar(fontsize = 15, fontface = "bold"))
#dev.copy2pdf(file = "internal_vs_external_table/Plots/fly_yeast_human_genegroups_internalvsexternal_significance_quadvenn_filtered.pdf")

```

## 19. Which genegroups have significant shifts in all datasets?

```{r}

# Create vector of genegroups which have significant shifts in all datasets (e.g. FDR < 0.2 for all datasets)
genegroups_significant_all_datasets <- (corr_analysis_wide %>% 
                                          filter(fly_fdr_GI < 0.2 & fdr_features < 0.2 & yeast_fdr_GI < 0.2 & human_fdr_GI < 0.2))[["genegroup"]] %>% sort()

length(genegroups_significant_all_datasets)
# [1] 32

print(genegroups_significant_all_datasets)
# [1] "ANAPHASE-PROMOTING COMPLEX"                        "AUTOPHAGY-RELATED GENES"                           "BRAHMA ASSOCIATED PROTEINS COMPLEX"               
# [4] "COAT PROTEIN COMPLEX II"                           "CONSERVED OLIGOMERIC GOLGI COMPLEX"                "COP9 SIGNALOSOME"                                 
# [7] "CYTOPLASMIC AMINOACYL-TRNA SYNTHETASES"            "CYTOPLASMIC LARGE RIBOSOMAL PROTEINS"              "CYTOPLASMIC SMALL RIBOSOMAL PROTEINS"             
#[10] "CYTOPLASMIC TRANSLATION INITIATION FACTORS"        "ESCRT-III COMPLEX"                                 "HEAT SHOCK PROTEIN 60 CHAPERONINS - GROUP II"     
#[13] "HISTONE PRE-MRNA CLEAVAGE COMPLEX"                 "MEDIATOR COMPLEX"                                  "MITOCHONDRIAL LARGE RIBOSOMAL PROTEINS"           
#[16] "NUCLEAR PORE COMPLEX"                              "OTHER GLYCOSYLTRANSFERASES"                        "PROTEASOME SUBUNITS"                              
#[19] "RNA POLYMERASE II"                                 "SPLICEOSOMAL SM PROTEINS"                          "SPLICEOSOME COMPLEX A"                            
#[22] "SPLICEOSOME COMPLEX B"                             "SPLICEOSOME COMPLEX C"                             "SPLICEOSOME COMPLEX P"                            
#[25] "SPT-ADA-GCN5 ACETYLTRANSFERASE"                    "TIP60 COMPLEX"                                     "TRANSCRIPTION FACTOR II D"                        
#[28] "U1 SMALL NUCLEAR RIBONUCLEOPROTEIN PARTICLE"       "U2 SMALL NUCLEAR RIBONUCLEOPROTEIN PARTICLE"       "U4-U6 SMALL NUCLEAR RIBONUCLEOPROTEIN PARTICLE"   
#[31] "U4-U6-U5 SMALL NUCLEAR RIBONUCLEOPROTEIN PARTICLE" "U5 SMALL NUCLEAR RIBONUCLEOPROTEIN PARTICLE"        

# Some of those seem familiar from the complex-annotated datasets and their internal vs. external shifts. Among those for example COP9 signalosome, mediator complex, TIP60 complex, RNAPolII, ESCRT, Proteasome

```

## 20. Jitter plots 

```{r}

# Define b110 theme:
theme_b110<-function(){
  theme_classic() +
  theme(
    axis.text=element_text(size = 16), 
    axis.title=element_text(size = 16),
    plot.title = element_text(size = 22,hjust = 0.5,face="bold"),
    legend.title = element_text(size = 22),
    legend.text = element_text(size =16),
    legend.position = "bottom"
    )
}

# In order to color dots according to correlation type, we need a new column type and the median values (of internal and external) need to be in the same column
# Create sub-datasets for internal and external (for GI and features):
corr_analysis_int <- corr_analysis %>% 
  select(organism, genegroup, median_r_GI_int) %>% 
  rename(median_r = median_r_GI_int) %>% 
  mutate(type = "internal") %>% 
  mutate(organism = ifelse(organism == "fly", "fly GI", organism))
corr_analysis_ext <- corr_analysis %>% 
  select(organism, genegroup, median_r_GI_ext) %>% 
  rename(median_r = median_r_GI_ext) %>% 
  mutate(type = "external") %>% 
  mutate(organism = ifelse(organism == "fly", "fly GI", organism))
corr_analysis_features_int <- corr_analysis %>% 
  select(organism, genegroup, median_r_features_int) %>% 
  rename(median_r = median_r_features_int) %>% 
  drop_na() %>% 
  mutate(type = "internal") %>% 
  mutate(organism = ifelse(organism == "fly", "fly m. f.", organism))
corr_analysis_features_ext <- corr_analysis %>% 
  select(organism, genegroup, median_r_features_ext) %>% 
  rename(median_r = median_r_features_ext) %>% 
  drop_na() %>% 
  mutate(type = "external") %>% 
  mutate(organism = ifelse(organism == "fly", "fly m. f.", organism))

# Create common, long-format dataframe:
corr_analysis_long <- rbind(corr_analysis_int, corr_analysis_ext, corr_analysis_features_int, corr_analysis_features_ext)

# Jitter plot without filtering for genegroups present in all datasets:
corr_analysis_long %>% 
ggplot(aes(x = organism, y = median_r, color = type, label = genegroup)) +
  theme_b110() +
  geom_jitter(width = 0.3, alpha = 0.4) +
  scale_color_manual(values = c("grey", "blue4")) +
  ylab("median pearson correlation") +
  xlab("dataset") +
  ggtitle(" Median pearson correlation values per gene group")
#ggsave(filename = "internal_vs_external_table/Plots/fly_yeast_human_genegroups_internalvsexternal_jitter.png", width = 9, height = 9)
#ggsave(filename = "internal_vs_external_table/Plots/fly_yeast_human_genegroups_internalvsexternal_jitter.pdf", width = 9, height = 9)

# Jitter plot after filtering for genegroups present in all datasets:
corr_analysis_long %>% 
  filter(genegroup %in% (corr_analysis_long %>% filter(organism == "yeast"))$genegroup) %>% 
  filter(genegroup %in% (corr_analysis_long %>% filter(organism == "human"))$genegroup) %>% 
  filter(genegroup %in% (corr_analysis_long %>% filter(organism == "fly GI"))$genegroup) %>% 
ggplot(aes(x = organism, y = median_r, color = type, label = genegroup)) +
  theme_b110() +
  geom_jitter(width = 0.3, alpha = 0.4) +
  scale_color_manual(values = c("grey60", "blue4")) +
  ylab("median pearson correlation") +
  xlab("dataset") +
  ggtitle(" Median pearson correlation values per gene group, \n filtered datasets")
#ggsave(filename = "internal_vs_external_table/Plots/fly_yeast_human_genegroups_internalvsexternal_jitter_filtered.png", width = 9, height = 9)
#ggsave(filename = "internal_vs_external_table/Plots/fly_yeast_human_genegroups_internalvsexternal_jitter_filtered.pdf", width = 9, height = 9)

```
